src/libsync/propagateuploadng.cpp

```diff
@@ -123,21 +250,39 @@ void PropagateUploadFileNG::slotPropfindIterate(const QString &name, const QMap<
     }
 }
 
+void PropagateUploadFileNG::updateRanges(quint64 start, quint64 size)
+{
+    for(auto iter = _serverRanges.begin(); iter != _serverRanges.end(); ++iter)
+    {
+        if (iter->start <= start && iter->size >= size) {
+            iter->start = start + size;
+            iter->size -= size;
+            if (iter->size == 0) {
+                _serverRanges.erase(iter);
+                break;
+            }
+        }
+    }
+}
+
 void PropagateUploadFileNG::slotPropfindFinished()
 {
     auto job = qobject_cast<LsColJob *>(sender());
     slotJobDestroyed(job); // remove it from the _jobs list
     propagator()->_activeJobList.removeOne(this);
 
-    _currentChunk = 0;
+    _currentChunk = _serverRanges.first().start;
     _sent = 0;
     while (_serverChunks.contains(_currentChunk)) {
```

Comment:
This logic seems unsafe to me. Say the server has chunks starting at 0-1000, 1000-2000 from a previous aborted no-zsync upload. Now we upload again with zsync and it reports the range 1500-1700. Things would very likely go very wrong.

We probably want a 'was zsync used' flag in the UploadInfo, so we never resume in a different way.

But even if you don't toggle that you can run into trouble here. Say the zsync metadata file on the server changed since the last zsync upload was aborted. Then we'd try to resume but the chunks needed could be totally different. That'd certainly not work. Maybe there are other cases.

Reply:
I just double checked this, and made `updateRanges` return a `bool` so removing a `_serverChunks` will be for anything that doesn't start at our the exact offset we expect and has a size smaller or equal to our delta-chunk.

```
11-03 15:10:48:857 [ info sync.networkjob.zsync.put ]:	Got remote zsync file for: "1g.rand.img" size: 8316
11-03 15:10:48:858 [ info sync.propagate.zsync.seed ]:	Reading seed file: "/tmp/test/1g.rand.img"
11-03 15:11:04:072 [ info sync.networkjob.zsync.put ]:	Done reading: "1g.rand.img" 69.3% of target seeded.
11-03 15:11:04:073 [ info sync.accessmanager ]:	6 "PROPFIND" "http://localhost/~aammar/owncloud/remote.php/dav/uploads/admin/512066796" has X-Request-ID "0c96d527-cab9-4398-bc86-8224fa20f92b"
11-03 15:11:04:073 [ info sync.networkjob ]:	OCC::LsColJob created for "http://localhost/~aammar/owncloud" + "" "OCC::PropagateUploadFileNG"
11-03 15:11:07:150 [ info sync.networkjob.lscol ]:	LSCOL of QUrl("http://localhost/~aammar/owncloud/remote.php/dav/uploads/admin/512066796") FINISHED WITH STATUS QNetworkReply::NetworkError(NoError) ""
11-03 15:11:07:150 [ info sync.propagator.upload ]:	Resuming  "1g.rand.img"  from chunk  128974848 ; sent = 0
11-03 15:11:07:150 [ info sync.propagator.upload ]:	To Delete (0, 10000000, 110000000, 210000000)
11-03 15:11:07:150 [ info sync.accessmanager ]:	5 "" "http://localhost/~aammar/owncloud/remote.php/dav/uploads/admin/512066796/0000000000000000" has X-Request-ID "2f013604-7200-45d6-9934-10cf4fca7349"
11-03 15:11:07:151 [ info sync.networkjob ]:	OCC::DeleteJob created for "http://localhost/~aammar/owncloud" + "" "OCC::PropagateUploadFileNG"
11-03 15:11:07:151 [ info sync.accessmanager ]:	5 "" "http://localhost/~aammar/owncloud/remote.php/dav/uploads/admin/512066796/0000000010000000" has X-Request-ID "cb1bf286-e5bb-45ad-a360-154820d3be92"
11-03 15:11:07:151 [ info sync.networkjob ]:	OCC::DeleteJob created for "http://localhost/~aammar/owncloud" + "" "OCC::PropagateUploadFileNG"
11-03 15:11:07:151 [ info sync.accessmanager ]:	5 "" "http://localhost/~aammar/owncloud/remote.php/dav/uploads/admin/512066796/0000000110000000" has X-Request-ID "d81aadef-546c-4738-8f3b-bf5f29681411"
11-03 15:11:07:151 [ info sync.networkjob ]:	OCC::DeleteJob created for "http://localhost/~aammar/owncloud" + "" "OCC::PropagateUploadFileNG"
11-03 15:11:07:151 [ info sync.accessmanager ]:	5 "" "http://localhost/~aammar/owncloud/remote.php/dav/uploads/admin/512066796/0000000210000000" has X-Request-ID "fc654e82-24b4-4608-a67b-10f80634cf5d"
11-03 15:11:07:152 [ info sync.networkjob ]:	OCC::DeleteJob created for "http://localhost/~aammar/owncloud" + "" "OCC::PropagateUploadFileNG"
11-03 15:11:10:305 [ info sync.networkjob.delete ]:	DELETE of QUrl("http://localhost/~aammar/owncloud/remote.php/dav/uploads/admin/512066796/0000000110000000") FINISHED WITH STATUS QNetworkReply::NetworkError(NoError) ""
11-03 15:11:10:416 [ info sync.networkjob.delete ]:	DELETE of QUrl("http://localhost/~aammar/owncloud/remote.php/dav/uploads/admin/512066796/0000000010000000") FINISHED WITH STATUS QNetworkReply::NetworkError(NoError) ""
11-03 15:11:10:525 [ info sync.networkjob.delete ]:	DELETE of QUrl("http://localhost/~aammar/owncloud/remote.php/dav/uploads/admin/512066796/0000000000000000") FINISHED WITH STATUS QNetworkReply::NetworkError(NoError) ""
11-03 15:11:10:639 [ info sync.networkjob.delete ]:	DELETE of QUrl("http://localhost/~aammar/owncloud/remote.php/dav/uploads/admin/512066796/0000000210000000") FINISHED WITH STATUS QNetworkReply::NetworkError(NoError) ""
11-03 15:11:10:643 [ info sync.accessmanager ]:	3 "" "http://localhost/~aammar/owncloud/remote.php/dav/uploads/admin/512066796/0000000128974848" has X-Request-ID "3ca4d2fc-16f5-498d-a928-5beb4373e329"
11-03 15:11:10:643 [ info sync.networkjob ]:	OCC::PUTFileJob created for "http://localhost/~aammar/owncloud" + "" "OCC::PropagateUploadFileNG"
11-03 15:11:13:789 [ info sync.networkjob.put ]:	PUT of "http://localhost/~aammar/owncloud/remote.php/dav/uploads/admin/512066796/0000000128974848" FINISHED WITH STATUS QNetworkReply::NetworkError(NoError) "" QVariant(int, 201) QVariant(QString, "Created")
11-03 15:11:13:790 [ info sync.propagator.upload ]:	Chunked upload of 10000000 bytes took 3147 ms, desired is 60000 ms, expected good chunk size is 190657769 bytes and nudged next chunk size to  100000000 bytes
11-03 15:11:13:857 [ info sync.accessmanager ]:	3 "" "http://localhost/~aammar/owncloud/remote.php/dav/uploads/admin/512066796/0000000138974848" has X-Request-ID "a910787a-3ecf-4a9a-9a75-a65e0541908b"
11-03 15:11:13:857 [ info sync.networkjob ]:	OCC::PUTFileJob created for "http://localhost/~aammar/owncloud" + "" "OCC::PropagateUploadFileNG"
11-03 15:11:17:818 [ info sync.networkjob.put ]:	PUT of "http://localhost/~aammar/owncloud/remote.php/dav/uploads/admin/512066796/0000000138974848" FINISHED WITH STATUS QNetworkReply::NetworkError(NoError) "" QVariant(int, 201) QVariant(QString, "Created")
11-03 15:11:17:818 [ info sync.propagator.upload ]:	Chunked upload of 94857600 bytes took 3962 ms, desired is 60000 ms, expected good chunk size is 1436510853 bytes and nudged next chunk size to  100000000 bytes
11-03 15:11:17:885 [ info sync.accessmanager ]:	3 "" "http://localhost/~aammar/owncloud/remote.php/dav/uploads/admin/512066796/0000000438304768" has X-Request-ID "cd8981fc-cb7a-447a-95b3-08d34621f801"
11-03 15:11:17:885 [ info sync.networkjob ]:	OCC::PUTFileJob created for "http://localhost/~aammar/owncloud" + "" "OCC::PropagateUploadFileNG"
11-03 15:11:21:968 [ info sync.networkjob.put ]:	PUT of "http://localhost/~aammar/owncloud/remote.php/dav/uploads/admin/512066796/0000000438304768" FINISHED WITH STATUS QNetworkReply::NetworkError(NoError) "" QVariant(int, 201) QVariant(QString, "Created")
11-03 15:11:21:968 [ info sync.propagator.upload ]:	Chunked upload of 100000000 bytes took 4083 ms, desired is 60000 ms, expected good chunk size is 1469507714 bytes and nudged next chunk size to  100000000 bytes
11-03 15:11:22:037 [ info sync.accessmanager ]:	3 "" "http://localhost/~aammar/owncloud/remote.php/dav/uploads/admin/512066796/0000000538304768" has X-Request-ID "13ab6d31-8186-48d8-9f86-09fd6fca6e55"
11-03 15:11:22:037 [ info sync.networkjob ]:	OCC::PUTFileJob created for "http://localhost/~aammar/owncloud" + "" "OCC::PropagateUploadFileNG"
11-03 15:11:25:120 [ debug sync.propagator.upload ]	[ OCC::PropagateUploadFileNG::slotZsyncGenerationFinished ]:	Finished generation of: "/var/folders/wl/jnlwt70x4r15qbr3tr4xvkw40000gp/T/ownCloud.J83452" size: 8316
11-03 15:11:25:121 [ debug sync.propagator.upload ]	[ OCC::PropagateUploadFileNG::slotZsyncGenerationFinished ]:	Starting upload of .zsync
11-03 15:11:25:121 [ info sync.accessmanager ]:	3 "" "http://localhost/~aammar/owncloud/remote.php/dav/uploads/admin/512066796/.zsync" has X-Request-ID "1b42f64b-a98d-4335-a927-768bdafc5ab4"
11-03 15:11:25:121 [ info sync.networkjob ]:	OCC::PUTFileJob created for "http://localhost/~aammar/owncloud" + "" "OCC::PropagateUploadFileNG"
11-03 15:11:26:268 [ info sync.networkjob.put ]:	PUT of "http://localhost/~aammar/owncloud/remote.php/dav/uploads/admin/512066796/0000000538304768" FINISHED WITH STATUS QNetworkReply::NetworkError(NoError) "" QVariant(int, 201) QVariant(QString, "Created")
11-03 15:11:26:268 [ info sync.propagator.upload ]:	Chunked upload of 100000000 bytes took 4231 ms, desired is 60000 ms, expected good chunk size is 1418104467 bytes and nudged next chunk size to  100000000 bytes
11-03 15:11:26:284 [ info sync.accessmanager ]:	3 "" "http://localhost/~aammar/owncloud/remote.php/dav/uploads/admin/512066796/0000000638304768" has X-Request-ID "f709450b-4cc6-46a9-9153-84bb57f0f2d9"
11-03 15:11:26:284 [ info sync.networkjob ]:	OCC::PUTFileJob created for "http://localhost/~aammar/owncloud" + "" "OCC::PropagateUploadFileNG"
11-03 15:11:28:203 [ info sync.networkjob.put ]:	PUT of "http://localhost/~aammar/owncloud/remote.php/dav/uploads/admin/512066796/.zsync" FINISHED WITH STATUS QNetworkReply::NetworkError(NoError) "" QVariant(int, 201) QVariant(QString, "Created")
11-03 15:11:28:203 [ debug sync.propagator.upload ]	[ OCC::PropagateUploadFileNG::doFinalMove ]:	Uploading of .zsync complete
11-03 15:11:28:203 [ info sync.accessmanager ]:	6 "MOVE" "http://localhost/~aammar/owncloud/remote.php/dav/uploads/admin/512066796/.file" has X-Request-ID "bb73d728-6b32-42cb-8a61-ac8c84c4f2b3"
11-03 15:11:28:203 [ info sync.networkjob ]:	OCC::MoveJob created for "http://localhost/~aammar/owncloud" + "" "OCC::PropagateUploadFileNG"
11-03 15:11:29:592 [ info sync.networkjob.put ]:	PUT of "http://localhost/~aammar/owncloud/remote.php/dav/uploads/admin/512066796/0000000638304768" FINISHED WITH STATUS QNetworkReply::NetworkError(NoError) "" QVariant(int, 201) QVariant(QString, "Created")
11-03 15:11:55:086 [ info sync.networkjob.move ]:	MOVE of QUrl("http://localhost/~aammar/owncloud/remote.php/dav/uploads/admin/512066796/.file") FINISHED WITH STATUS QNetworkReply::NetworkError(NoError) ""
11-03 15:11:55:086 [ debug sync.propagator.upload ]	[ OCC::PropagateUploadFileNG::slotMoveJobFinished ]:	*==* duration UPLOAD 1073741824 2633 0 66842
11-03 15:11:55:086 [ info sync.database ]:	Updating file record for path: "1g.rand.img" inode: 8603287640 modtime: 1509714432 type: 0 etag: "803a7a279288e9dd5a580696ba13352a" fileId: "00007254oc1p5kwdqwsc" remotePerm: "WDNVR" fileSize: 1073741824 checksum: "SHA1:40d04b995234595a1415cca31f731140fb0c9610"
11-03 15:11:55:087 [ info sync.propagator ]:	Completed propagation of "1g.rand.img" by OCC::PropagateUploadFileNG(0x7ffbc9e54bd0) with status 4
11-03 15:11:55:087 [ info sync.database ]:	Closing DB "/tmp/test/._sync_3eda4b7f1fe3.db"
11-03 15:11:55:089 [ info sync.engine ]:	CSync run took  72587 ms
11-03 15:11:55:090 [ info sync.database ]:	Closing DB "/tmp/test/._sync_3eda4b7f1fe3.db"
```

Reply:
This still sounds like a problem when the zsync metadata file changes before the resuming upload. Say the initial upload thinks two chunks need to be uploaded: 1000-2000 and 3000-4000. Then when resuming it turns out we only need to upload chunk 0-1000. The check _serverChunks.contains(0) would fail, we'd just upload 0-1000, then call move but the old data would stay on the server!

Reply:
Will come up with something better, I see the issue now.

Reply:
Done.